diff --git a/conf/env.mk b/conf/env.mk
index a603f9e..2502dd5 100644
--- a/conf/env.mk
+++ b/conf/env.mk
@@ -16,5 +16,4 @@ V = @
 
 # If the makefile cannot find your QEMU binary, uncomment the
 # following line and set it to the full path to QEMU.
-#
-# QEMU=
+QEMU=/home/skoralah/qemu/bin/qemu-system-i386
diff --git a/kern/kdebug.c b/kern/kdebug.c
index 9547143..5653aa2 100644
--- a/kern/kdebug.c
+++ b/kern/kdebug.c
@@ -179,7 +179,13 @@ debuginfo_eip(uintptr_t addr, struct Eipdebuginfo *info)
 	//	Look at the STABS documentation and <inc/stab.h> to find
 	//	which one.
 	// Your code here.
-
+ 
+//	If *region_left > *region_right, then 'addr' is not contained in any
+//	matching stab.
+		stab_binsearch(stabs, &lline, &rline, N_SLINE, addr);
+                 if(lline > rline)
+                 return -1;
+		 info->eip_line = stabs[rline].n_desc;
 
 	// Search backwards from the line number for the relevant filename
 	// stab.
diff --git a/kern/monitor.c b/kern/monitor.c
index e137e92..77c4ec2 100644
--- a/kern/monitor.c
+++ b/kern/monitor.c
@@ -24,6 +24,7 @@ struct Command {
 static struct Command commands[] = {
 	{ "help", "Display this list of commands", mon_help },
 	{ "kerninfo", "Display information about the kernel", mon_kerninfo },
+	{ "backtrace", "Display function backtrace", mon_backtrace },
 };
 
 /***** Implementations of basic kernel monitor commands *****/
@@ -57,7 +58,16 @@ mon_kerninfo(int argc, char **argv, struct Trapframe *tf)
 int
 mon_backtrace(int argc, char **argv, struct Trapframe *tf)
 {
-	// Your code here.
+	uint32_t* p=(uint32_t *) read_ebp();
+	cprintf("Stack backtrace:\n");
+	while(p)
+	{
+	 	struct Eipdebuginfo info;
+		cprintf("ebp %08x eip %08x args %08x %08x %08x %08x %08x\n",p, *(p+1),*(p+2),*(p+3),*(p+4),*(p+5),*(p+6));
+		debuginfo_eip(*(p+1), &info);
+		cprintf("\t%s:%d : %.*s+%u\n",info.eip_file, info.eip_line, info.eip_fn_namelen, info.eip_fn_name, (*(p+1)-info.eip_fn_addr));
+		p=(uint32_t*)*p;
+	}
 	return 0;
 }
 
